api/src/main/java/org/pmiops/workbench/cohortreview/AnnotationQueryBuilder.java

```diff
@@ -0,0 +1,219 @@
+import org.pmiops.workbench.db.dao.CohortAnnotationDefinitionDao;
+import org.pmiops.workbench.db.model.CohortAnnotationDefinition;
+import org.pmiops.workbench.db.model.CohortReview;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.model.AnnotationQuery;
+import org.pmiops.workbench.model.AnnotationType;
+import org.pmiops.workbench.model.CohortStatus;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Service;
+
+@Service
+public class AnnotationQueryBuilder {
+
+  public static final String PERSON_ID_COLUMN = "person_id";
+  public static final String REVIEW_STATUS_COLUMN = "review_status";
+
+  // These column names are reserved and can't be used for annotation definition column names.
+  public static final ImmutableSet<String> RESERVED_COLUMNS = ImmutableSet.of(PERSON_ID_COLUMN,
+      REVIEW_STATUS_COLUMN);
+
+  private static final ImmutableMap<AnnotationType, String> ANNOTATION_COLUMN_MAP =
+      ImmutableMap.of(
+          AnnotationType.BOOLEAN, "annotation_value_boolean",
+          AnnotationType.DATE, "annotation_value_date",
+          AnnotationType.INTEGER, "annotation_value_integer",
+          AnnotationType.STRING, "annotation_value_string");
+
+  private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+  private final CohortAnnotationDefinitionDao cohortAnnotationDefinitionDao;
+
+  @Autowired
+  AnnotationQueryBuilder(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
+      CohortAnnotationDefinitionDao cohortAnnotationDefinitionDao) {
+    this.namedParameterJdbcTemplate = namedParameterJdbcTemplate;
+    this.cohortAnnotationDefinitionDao = cohortAnnotationDefinitionDao;
+  }
+
+  private static final String ANNOTATION_JOIN_SQL =
+      " LEFT OUTER JOIN participant_cohort_annotations a%d "
+      + "ON a%d.participant_id = pcs.participant_id "
+      + "AND a%d.cohort_annotation_definition_id = %d"
+      + "AND a%d.cohort_review_id = %d";
+
+  private static final String ANNOTATION_VALUE_JOIN_SQL =
+      " LEFT OUTER JOIN cohort_annotation_enum_value ae%d "
+      + "ON ae%d.cohort_annotation_enum_value_id = a%d.cohort_annotation_enum_value_id";
+
+  private static final String WHERE_SQL =
+      " WHERE pcs.cohort_review_id = %d";
+
+  private static final ImmutableSet<CohortStatus> REVIEWED_STATUSES =
+      ImmutableSet.of(CohortStatus.INCLUDED, CohortStatus.EXCLUDED, CohortStatus.NEEDS_FURTHER_REVIEW);
+
+  private Map<String, CohortAnnotationDefinition> getAnnotationDefinitions(CohortReview cohortReview) {
+    return Maps.uniqueIndex(
+            cohortAnnotationDefinitionDao.findByCohortId(cohortReview.getCohortId()),
+            CohortAnnotationDefinition::getColumnName);
+  }
+
+  private String getSelectAndFromSql(CohortReview cohortReview, List<String> columns,
```

Comment:
I'm not particularly concerned with SQL string building against BigQuery. It does seem substantially riskier to be doing this against MySQL though. Is there an open source library available to enforce building a safe SQL string?

Reply:
None of the values being passed up in the request are being used in the generated SQL statements, so there should be no risk of SQL injection here. The column names in the request are either person_id, review_status, or an annotation name that we use to look up a definition in the annotationDefinitions map (but is not passed into SQL anywhere.)

Sorry if that's less than clear. :)

Reply:
I see. Yeah it's very difficult to verify this kind of thing by eye (and in practice, the more realistic concern is someone less savvy coming in and adding a small feature and introducing a security issue).

Setting aside whether there are any security issues in this current PR, my question still stands on whether there's a library we can use to build a safe SQL string instead.